### PR TITLE
8298126: Print statistics for objects in CDS archive heap

### PR DESCRIPTION
**Pull request by iklam**
For diagnostic purposes, it's useful to print out statistics of the objects in the CDS archive heap. For example, this shows the number of objects with a distribution of different sizes.

```
[6.772s][info][cds,heap]        0 objects are <= 8      bytes (total        0 bytes, avg      0.0 bytes)
[6.772s][info][cds,heap]     1205 objects are <= 16     bytes (total    19280 bytes, avg     16.0 bytes)
[6.772s][info][cds,heap]    15717 objects are <= 32     bytes (total   411024 bytes, avg     26.2 bytes)
[6.772s][info][cds,heap]     4127 objects are <= 64     bytes (total   198080 bytes, avg     48.0 bytes)
[6.772s][info][cds,heap]     2380 objects are <= 128    bytes (total   243016 bytes, avg    102.1 bytes)
[6.772s][info][cds,heap]      293 objects are <= 256    bytes (total    47920 bytes, avg    163.5 bytes)
[6.772s][info][cds,heap]       51 objects are <= 512    bytes (total    17112 bytes, avg    335.5 bytes)
[6.772s][info][cds,heap]       18 objects are <= 1024   bytes (total    14480 bytes, avg    804.4 bytes)
[6.772s][info][cds,heap]       11 objects are <= 2048   bytes (total    13440 bytes, avg   1221.8 bytes)
[6.772s][info][cds,heap]        3 objects are <= 4096   bytes (total     9808 bytes, avg   3269.3 bytes)
[6.772s][info][cds,heap]        3 objects are <= 8192   bytes (total    12840 bytes, avg   4280.0 bytes)
[6.772s][info][cds,heap]        4 objects are <= 16384  bytes (total    37240 bytes, avg   9310.0 bytes)
[6.772s][info][cds,heap]        0 objects are <= 32768  bytes (total        0 bytes, avg      0.0 bytes)
[6.772s][info][cds,heap]        0 objects are <= 65536  bytes (total        0 bytes, avg      0.0 bytes)
[6.772s][info][cds,heap]        0 objects are <= 131072 bytes (total        0 bytes, avg      0.0 bytes)
[6.772s][info][cds,heap]        0 objects are <= 262144 bytes (total        0 bytes, avg      0.0 bytes)
[6.772s][info][cds,heap]        0 huge  objects               (total        0 bytes, avg      0.0 bytes)
[6.772s][info][cds,heap]    23812 total objects               (total  1024240 bytes, avg     43.0 bytes)
```

These stats are useful when implement large changes in the CDS heap archiving code. For example, [JDK-8296344](https://bugs.openjdk.org/browse/JDK-8296344) significantly changes how the objects are copied into the archive. We can use these stats to verify that the exact same objects are copied.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8298126](https://bugs.openjdk.org/browse/JDK-8298126): Print statistics for objects in CDS archive heap

### Reviewers
 * @ashu-mehra (no known github.com user name / role) ⚠️ Review applies to [9c48fd09](https://git.openjdk.org/jdk/pull/11520/files/9c48fd0992c4073c49082dd7326793825c579af6)
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to [9c48fd09](https://git.openjdk.org/jdk/pull/11520/files/9c48fd0992c4073c49082dd7326793825c579af6)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11520/head:pull/11520` \
`$ git checkout pull/11520`

Update a local copy of the PR: \
`$ git checkout pull/11520` \
`$ git pull https://git.openjdk.org/jdk pull/11520/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11520`

View PR using the GUI difftool: \
`$ git pr show -t 11520`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11520.diff">https://git.openjdk.org/jdk/pull/11520.diff</a>

</details>

### CODE REVIEW

**Comment by ashu-mehra** [comment]
@iklam does it make sense to collect these stats only for non-product builds as these are not intended for end-users.

File: src/hotspot/share/cds/heapShared.cpp [hunk 1]
```diff
@@ -831,6 +838,7 @@ void HeapShared::write_subgraph_info_table() {
     _archived_ArchiveHeapTestClass = array;
   }
 #endif
+  print_stats();
```
Comment (ashu-mehra) [inline src/hotspot/share/cds/heapShared.cpp:841]: Perhaps make it conditional on `log_is_enabled(Info, cds, heap)`.

**Comment by iklam** [comment]
> @iklam does it make sense to collect these stats only for non-product builds as these are not intended for end-users.

We usually leave the stats in the product build if they are not in a performance critical path. That's true in general for the logging code. You can never have too much logging :-)

File: src/hotspot/share/cds/heapShared.cpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/cds/heapShared.cpp]: Fixed as suggested.

**Review by ashu-mehra — APPROVED** [review]

**Review by calvinccheung — APPROVED** [review]
LGTM

**Comment by iklam** [comment]
Thanks @ashu-mehra and @calvinccheung for the review.

**Pull request closed without merging**